src/molinillo_solver.cr

```diff
@@ -47,7 +47,36 @@ module Shards
       end
     end
 
+    private def prefetch_local_caches(deps)
+      count = 0
+      active = 0
+      ch = Channel(Exception?).new(deps.size + 1)
+      deps.each do |dep|
+        count += 1
+        active += 1
+        while active > Shards.parallel_fetch
+          sleep 0.1
+        end
+        spawn do
+          begin
+            dep.resolver.update_local_cache
+            ch.send(nil)
+            active -= 1
+          rescue ex : Exception
+            ch.send(ex)
+          end
+        end
+      end
+
+      count.times do
+        obj = ch.receive
+        raise obj if obj.is_a? Exception
+      end
+    end
+
     private def add_lock(base, lock_index, deps : Array(Dependency))
+      prefetch_local_caches(deps) if Shards.parallel_fetch > 1
```

Comment:
Using early return within `prefetch_local_caches` instead of in the callsite would DRY things up a bit.

```cr
return unless Shards.parallel_fetch > 1
```

Reply:
Good point, fixed in https://github.com/crystal-lang/shards/pull/539/commits/f3a30698ccb71c0f796d4931413f35c518e3eeb0